subsys/bluetooth/controller/ll_sw/ull_iso.c

```diff
@@ -1280,26 +1334,12 @@ int ull_iso_reset(void)
 #if defined(CONFIG_BT_CTLR_ADV_ISO) || defined(CONFIG_BT_CTLR_CONN_ISO)
 void ull_iso_lll_ack_enqueue(uint16_t handle, struct node_tx_iso *node_tx)
 {
-	struct ll_iso_datapath *dp = NULL;
-
-	if (IS_ENABLED(CONFIG_BT_CTLR_CONN_ISO) && IS_CIS_HANDLE(handle)) {
-		struct ll_conn_iso_stream *cis;
-
-		cis = ll_conn_iso_stream_get(handle);
-		dp  = cis->hdr.datapath_in;
-
-		if (dp) {
-			isoal_tx_pdu_release(dp->source_hdl, node_tx);
-		}
-	} else if (IS_ENABLED(CONFIG_BT_CTLR_ADV_ISO) && IS_ADV_ISO_HANDLE(handle)) {
-		/* Process as TX ack. TODO: Can be unified with CIS and use
-		 * ISOAL.
-		 */
-		ll_tx_ack_put(handle, (void *)node_tx);
-		ll_rx_sched();
-	} else {
-		LL_ASSERT(0);
-	}
+	/* ISO Tx ack, we are in LLL context */
+	/* FIXME: add new interface to pass through ULL context as
+	 * ll_tx_ack_put shall only be ULL callable.
+	 */
+	ll_tx_ack_put(handle, (void *)node_tx);
+	ll_rx_sched();
```

Comment:
The idea is that we call `isoal_tx_pdu_release` and ISO-AL uses its pdu_release implementation for the datapath. For HCI this means `ll_iso_pdu_release` is called, which does the ll_tx_ack_get etc. Please restore.

Reply:
`ull_iso_lll_ack_enqueue()` is called from LLL execution context, it is incorrect to call `isoal_tx_pdu_release()` directly from LLL execution context, this will corrupt ISOAL where the alloc/write/emit execute in Thread execution context (and I assume for vendor datapath, in ULL execution context) . For vendor data path, `isoal_tx_pdu_release()` will need to be called in the (ULL?) execution context where the alloc/write/emit are executed in order to respect context safety.

Reply:
In the implementation of `isoal_tx_pdu_release` you will see that it calls a data path-specific release function. For HCI path it calls `ll_iso_pdu_release`, which does exactly what you want: `ll_tx_ack_put`/`sched`.
For vendor path, another release function is called, which ensures context is correct.
Your change has no impact on CIS functionality for Nordic, and destroys it for vendors.